tartufo/scanner.py

```diff
@@ -543,11 +542,23 @@ def _iter_diff_index(
             if delta.is_binary:
                 self.logger.debug("Binary file skipped: %s", file_path)
                 continue
+            if delta.status == pygit2.GIT_DELTA_DELETED:
+                self.logger.debug("Skipping as the file is deleted")
+                continue
             printable_diff: str = patch.text
+            header_length = GitScanner.header_length(printable_diff)
+            printable_diff = printable_diff[header_length:]
             if self.should_scan(file_path):
                 # The `printable_diff` contains the full 4-line diff header,
                 # so we need to strip that before analyzing it
-                yield printable_diff.split("\n", 4)[4], file_path
+                yield printable_diff, file_path
+
+    @staticmethod
+    def header_length(diff: str) -> int:
+        """Compute the length of the git diff header text"""
+        b_file_pos = diff.find("+++")
```

Comment:
Color me paranoid, but I think we want this:
```suggestion
        b_file_pos = diff.find("\n+++") + 4
```

Reply:
I understand (and agree with the `\n`, but why `+ 4`?

Reply:
Well, if we do nothing, then `b_file_pos` points at the first character of the find, namely `\n` (instead of the first `+` as previously). Therefore the next find will match immediately, and the current position (`+++` followed by whatever) will be returned as the first part of the diff payload. In actuality, we know the newline we're seeking can't come any sooner than immediately following `+++`, i.e. 4 characters after the `\n` we matched.

Reply:
Got it. Okay, that makes sense!

Reply:
I think we need to add a comment here to explain the rationale (what @rbailey-godaddy said above) for using the `+ 4`

Reply:
This might fail in a specific scenario where the file is a rename with similarity index being 100%. The printable diff in this case would look something as follows:

```
diff --git a/.github/actions/wss-unified-agent.config b/.github/actions/run-wss-scan/wss-unified-agent.config
similarity index 100%
rename from .github/actions/wss-unified-agent.config
rename to .github/actions/run-wss-scan/wss-unified-agent.config
```

In this case there will not be a `+++`, so it will end up truncating the first line in the above diff. I believe it should be okay. Just wanted to get it to your notice before I make changes

Reply:
Interesting, then maybe we have something like

```suggestion
        try:
            b_file_pos = diff.find("\n+++") + 4  # First character after the +++
            final_nl = diff.find("\n", b_file_pos)
        except ValueError:
            final_nl = -1  # This will return 0, since we add 1 at the end
```

Reply:
It will not raise an error. Instead diff.find("\n+++") will return -1. So going with the following code snippet if it looks okay?

```
@staticmethod
    def header_length(diff: str) -> int:
        """Compute the length of the git diff header text"""
        b_file_pos = diff.find("\n+++")
        if b_file_pos != -1:
            b_file_pos = b_file_pos + 4  # Retrieving first character after the +++
        final_nl = diff.find("\n", b_file_pos)
        return final_nl + 1
```
Retrieving exact position if there is an occurrence of "\n+++"

Reply:
Ah, right! I was getting confused between `.find()` and `.index()`. Using `.find()` returns `-1`, where `.index()` will raise a `ValueError` 😄 
Looks great!

Reply:
I still can't follow this. What about:

```
@staticmethod
def header_length(diff: str) -> int:
    """Compute the length of the git diff header text"""
    try:
        # Header ends after newline following line starting with "+++"
        b_file_pos = diff.index("\n+++")
        return diff.index("\n", b_file_pos + 4) + 1
    else:
        # If no separator, diff is header only (pure rename)
        return len(diff)
```

Reply:
Sure this looks good and we can have this. 
By returning len(diff) in except block we are not going to process the header at all in case of pure renames.